### PR TITLE
200m Nearby Scanning

### PR DESCRIPTION
**Pull request by khag7**
## Description
When scanning at a step distance of 200m a MUCH larger area can be covered. Existing database information (spawnpoint id and encounter id's) can be used to pinpoint a spawnpoint (lat, lng, and time) given only the encounter id from the "nearby pokemons" list.

## Motivation and Context
Faster scanning.

A big part of what needs to be done to make this work is to come up with a function to pull all encounter_id's for a given spawnpoint from the db and to calculate what the expected time-based bits in the encounter_id should be right now. These can then be compared to the actual bits in the encounter_id for the nearby pokemon to determine if its a match for that spawnpoint or not.

To be even more efficient (by making less db queries) we would need to add a spawnpoints table to the db to store the encounter bit data.

## How Has This Been Tested?
Hasn't yet, it's just a very basic plan, no code yet.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by 4RK4N** [comment]
so if i get this right: u basically want to grab the data the ingame "sightings" menu gets and then check that with the database to narrow down the exact spawnpoint ? in theory sounds good. was thinking about smth like that as well. but my python skills and my free time are too low to dive into this project to make a PR.

**Comment by ottowayne** [comment]
> Hasn't yet, it's just a very basic plan, no code yet.

Why is this a pull request then?

**Comment by jchristi** [comment]
I remember a thread on Reddit which was trying to extract information from the encounter id. Is it possible to get that information? That part I don't know but if anyone does I would be happy to implement this idea.

**Comment by khag7** [comment]
@ottowayne because this is a community where people work on things together

its much easier for me to post this BEFORE i start spending a lot of time on it, so people can propose ideas  and suggestions NOW instead of later.

Also, I'm not the most skilled python coder, surely someone can implement this in a working manner faster than I can.

It's a pull request in progress

**Comment by khag7** [comment]
@4RK4N @jchristi there is a reddit post which describes how to use the history of encounter id's at a specific spawnpoint to determine if an unknown encounter id could possibly have come from that spawnpoint:
https://www.reddit.com/r/pokemongodev/comments/4z9a6d/encounter_ids_not_completely_random_some_bits/

**Comment by ottowayne** [comment]
@khag7 It's just that feature requests/suggestions should be in issues. If you don't plan on implementing it, then don't make a pull request

**Comment by jchristi** [comment]
@ottowayne there is psuedo code which is a start. With the provided information, I think its enough that I can implement and I think WIP PRs are totally acceptable given the collaboration it enables between @khag7 and I.

@khag7 yes, that's the post I remember seeing though at the time it seemed incomplete but I will read over it in more detail when I get back in town. Thanks.

**Comment by khag7** [comment]
Good news, bad news:
I am very far along in the code needed to make this happen.
I am very certain it is probably inefficient and someone smarter than I can clean it up.

The code I have does not work, but I know its most of the way there. I'm gonna get my branch updated today and someone hopefully can pick it apart and find out what needs done to make it work.

**Comment by jchristi** [comment]
@khag7 what is it specifically that isn't working?

**Comment by khag7** [comment]
image

i know this doesn't look like much, but what you are seeing is detection of pokemon from 200m away. you can see clearly where i scanned (to the left and right) yet i found pokemon in between, and I know their encounter_id and what time they will expire. need to clean up the code a bit and fix a ton of merge conflicts but its working!

**Comment by ottowayne** [comment]
Nice to see.
I think there is some potential in this. For example if you only want to find a few rare pokemon you could define them and if your approximate search finds a selected pokemon, it could check all spawnpoints in a 200m radius that have a spawntime of t-15m to t-0.
If you think even further ahead it would be possible to build a cluster generation script like #990 that merges spawnpoints to 200m radius clusters.
I already thought about stuff like this but didn't have the time and motivation to implement this. Maybe you are more successful than me :+1:

**Comment by khag7** [comment]
@ottowayne well i've been coding python for about 5 weeks now so i'm not too efficient (takes me a long time to get things working, i make a ton of stupid syntax errors) but I've had the logic behind this conceptualized in my head for weeks now, i'm just struggling to make it work. I'm really hoping some python wizard jumps in here and takes my code and makes it bitchin' fast.

the current problem to solve is with the function which analyzes all old encounter_id's and generates the bit values for a sequence. either i have some bad data in my database, or the function is not working properly, or the theory behind all this is wrong. I think the data is fine, and i think the theory is fine. Something is wrong with my function. I'll keep trying at it but if anyone wants to get involved start reading through nearbyPokemonAnalysis.py and seeing how it works

**Comment by averylarry** [comment]
#990 will already accommodate a 200m radius.

**Comment by averylarry** [comment]
I guess I'm not sure if you are using the disappear time recorded in the database, but it should be noted that it is not accurate for a small percentage of spawnpoints (the ones that don't spawn once per hour for 15 minutes).

**Comment by khag7** [comment]
@averylarry funny you should mention that, I just came here to talk about that

for a small percentage of spawnpoints, the data is getting really f'd up when I try to determine what the bitvalue sequence should be. after picking apart the logs, I realized its only happening for spawnpoints which more than once per hr AND spawnpoints with inconsistent disappear times.

part of the problem is this:
lets say a spawnpoint has a disappear time of 12:10 AM (ten minutes past the hour). And lets say it happens to be one of those spawnpoints that spawns the same mon twice, 30 mins apart. The real way to consider this spawnpoint is as a 45 minute point with a 15 minute hiding period in the middle. So the mon spawns at HH:25 (in the previous hour). It gets observed by our scanner and we log it as a 15 minute spawn with a despawn of 11:40 PM (the previous day). If we don't happen to scan it again later, the encounter_id never gets updated with the new 12:10 AM on the next day. Then when my function tries to calculate the bitvalues based on the day and the hour, it chokes because we recorded the WRONG EXPIRATION TIME.

So really... for this to ever work well we have to start recording the proper expiration time. That involves doing research on each spawnpoint. By research I mean: visit the same point every 10 minutes for an hour or so. See when its there, when its not, when the timestamp is valid, when its not. The acquired information can be used to work out what type of spawnpoint it is. Then we have to store that in the db  so that every time this point gets scanned in the future, we can log it properly (correct despawn time). THEN finally this method can work because we can match the encounter_id bitvalues to the proper position in the day/hr based cycle/sequence

confused yet? I am.

**Comment by averylarry** [comment]
That's the reason why I think we need to add a table for spawnpoints.  The location, disappear time, and type of spawn are constant.
Of course, getting this merged is difficult enough.  A database schema change would make it even harder.

But how I would do it:
New spawnpoints get added to the spawnpoint table with id, location, type=0 and disappear time if given or disappear time of 3000-01-01 (something arbitrary, I hate null/none).
The pokemon table (which is really an encounter table) would no longer have location.  You would join the pokemon table with the new spawnpoint table for that information.  It would still have disappear time since this is the only good place to collect the data needed to determine the type of spawnpoint.  Well, maybe encounter table or something else (see below).

And then, of course, you have to run a process (hourly? daily?) that would look at the spawnpoints of unknown type and try to determine from the data in the pokemon table (and scannedlocation?) what the type is.  But at some point this requires additional scanning of locations, which is even more work and those scans have to be recorded somehow because we have to know that a spawnpoint was scanned but returned nothing.

That's off the top of my head and I have to go to work now.

**Comment by jchristi** [comment]
Is there at least a way to detect when the data gets screwed up so you can pass on those spawns? Scanning 200m and getting results for 90% (or whatever the percentage is) is way better than 0%. Getting that code merged would allow time and make it easier to get future improvements accepted.

**Comment by averylarry** [comment]
Exclude any spawnpoints that have more than 1 disappear minute:second (ignore the hour).

So basically return everything in the pokemon table and group by spawnpoint_id, minute:second.  THEN you remove any spawnpoints listed more than once.

**Comment by khag7** [comment]
@jchristi your suggestion is my plan. basically anything that has conflicting data will just be ignored. we'll get SOMETHING working for now, then try to fix the rest later.

**Comment by jorgebsa** [comment]
The developer from [fastpokemap](https://www.reddit.com/r/pokemongodev/comments/52q4y1/implementation_no_presentation_needed/) has supposedly found a way to do 200m scan, and if you check his website it does indeed seem to be running a working 200m scan. Maybe trying to reach out to him and validating a thing or two, not related to pyhton but how his logic works, could help you out

**Comment by j16sdiz** [comment]
See also: https://www.reddit.com/r/pokemongodev/comments/4z9a6d/encounter_ids_not_completely_random_some_bits/d7nnc5g

File: pogom/models.py [hunk 1]
```diff
@@ -571,36 +572,86 @@ def hex_bounds(center, steps):
...
-                    # Set a value of 15 minutes because currently its unknown but larger than 15.
-                    d_t = datetime.utcfromtimestamp((p['last_modified_timestamp_ms'] + 900000) / 1000.0)
 
-                printPokemon(p['pokemon_data']['pokemon_id'], p['latitude'],
-                             p['longitude'], d_t)
+            # add all wild pokes to the seen_pokemons list
+            seen_pokemons = cell.get('wild_pokemons', [])
+
+            # for testing nearby analysis -- ignore actual wild pokemons within 70m
+            seen_pokemons = []
```
Comment (jchristi) [inline pogom/models.py]: Obviously, you need to remove this for my suggestion below to work

File: pogom/models.py [hunk 2]
```diff
@@ -571,36 +572,86 @@ def hex_bounds(center, steps):
...
+            
+            # check to see if we're on the last cell. if so, analyze all the nearbys seen at this spawnpoint
+            cellcount += 1
+            if cellcount == len( cells ):
+                if len( nearby_pokemon ) > 0:
+                    log.info( 'found {} nearby pokemon, will try to analyze'.format( len( nearby_pokemon) ) )
+                    attempts += len( nearby_pokemon )
+                    try:
+                        reload( nearbyPokemonsAnalysis )
+                        nearby_analyzed = nearbyPokemonsAnalysis.analyze_nearby_pokemons(step_location,nearby_pokemon,use_extra_error_checking)
```
Comment (jchristi) [inline pogom/models.py]: pass `seen_pokemons` to `analyze_nearby_pokemons()`

**Comment by primaxius** [comment]
WIP do not belong in PR's

**Comment by ottowayne** [comment]
@primaxius You are discouraging developers to participate in the active development by doing this.
Just introduce a WIP tag and allow them as long as there is active development going on.

**Comment by jchristi** [comment]
@khag7 message me on Discord if you'd like to continue collaborating on this.

**Pull request closed without merging**